packages/solid-query/src/createBaseQuery.ts

```diff
@@ -35,46 +42,107 @@ export function createBaseQuery<
   const queryClient = createMemo(() =>
     useQueryClient({ context: options().context }),
   )
+
   const emptyData = Symbol('empty')
 
+  let resolver: (value: TData | undefined) => void
+  let queryResolver: (value: DehydratedState) => void
+
+  // This resource will be used in a Server environment to
+  // dehydrate the queryClient when the query is
+  // pre fetched on the server. It will always be undefined
+  // if an observer is mounted on client.
+  const [queryResource] = createResource<DehydratedState | undefined>(() => {
+    return new Promise((resolve) => {
+      if (!isServer) {
+        resolve(undefined)
+      }
+      queryResolver = resolve
+    })
+  })
+
+  // If queryResource is defined,
+  // This means that the query was fetched on the server!
+  // We need to hydrate the queryClient with the query from
+  // the server before we set up the observer and its results.
+  if (!isServer && queryResource()) {
+    hydrate(queryClient(), queryResource())
+  }
+
   const defaultedOptions = queryClient().defaultQueryOptions(options())
   defaultedOptions._optimisticResults = 'optimistic'
+  if (!isServer && queryResource()) {
+    defaultedOptions.refetchOnMount = false
+  }
   const observer = new Observer(queryClient(), defaultedOptions)
 
   const [state, setState] = createStore<QueryObserverResult<TData, TError>>(
     observer.getOptimisticResult(defaultedOptions),
   )
 
-  const [dataResource, { refetch, mutate }] = createResource<TData | undefined>(
+  const [dataResource, { mutate, refetch }] = createResource<TData | undefined>(
     () => {
       return new Promise((resolve) => {
-        if (!(state.isFetching && state.isLoading)) {
-          if ((unwrap(state.data) as TData | typeof emptyData) === emptyData) {
-            resolve(undefined)
+        if (isServer) {
+          if (!state.isInitialLoading) {
+            const dehydratedClient = dehydrate(queryClient())
+            queryResolver?.(dehydratedClient)
+            resolve(unwrap(state.data))
+          }
+          // We only resolve the data resource
+          // when the query observer finds a result
+          // This function will be called inside the observer
+          // subscription function.
+          resolver = resolve
+        } else {
+          if (!state.isInitialLoading) {
+            if (
+              (unwrap(state.data) as TData | typeof emptyData) === emptyData
+            ) {
+              resolve(undefined)
+            }
+            resolve(unwrap(state.data))
           }
-          resolve(unwrap(state.data))
         }
       })
     },
+    {
+      get deferStream() {
+        return options().deferStream
+      },
+    },
   )
 
-  batch(() => {
-    mutate(() => unwrap(state.data))
-    refetch()
-  })
+  if (!isServer) {
+    batch(() => {
+      mutate(() => unwrap(state.data))
+      refetch()
+    })
```

Comment:
After some experimenting, I believe this is at least one of the causes of the hydration key error - attempting to avoid this in #4840, but atm my alternative approach is causing most of the `expect(states.length).toBe(3)` assertions in the tests to be one larger than before (e.g. `expect(states.length).toBe(4)`). Haven't had time to dig into the tests quite yet.